src/track/beats.cpp

```diff
@@ -1,21 +1,158 @@
+        : Beats(track, iSampleRate) {
+    readByteArray(byteArray);
+}
+
+Beats::Beats(const Track* track, SINT iSampleRate)
+        : m_mutex(QMutex::Recursive),
+          m_track(track),
+          m_iSampleRate(iSampleRate == 0 ? m_track->getSampleRate() : iSampleRate),
+          m_dCachedBpm(0),
+          m_dLastFrame(0) {
+    // TODO(JVC) iSampleRate == 0 creates problems. Apparently only in tests
+
+    // BeatMap should live in the same thread as the track it is associated
+    // with.
+    moveToThread(track->thread());
+}
+
+// TODO(JVC) Do we really need a copy constructor??
+Beats::Beats(const Beats& other)
+        : m_mutex(QMutex::Recursive),
+          m_track(other.m_track),
+          m_subVersion(other.m_subVersion),
+          m_iSampleRate(other.m_iSampleRate),
+          m_dCachedBpm(other.m_dCachedBpm),
+          m_dLastFrame(other.m_dLastFrame),
+          m_beats(other.m_beats) {
+    moveToThread(m_track->thread());
+}
+
+void Beats::createFromBeatVector(const QVector<double>& beats) {
+    if (beats.isEmpty()) {
+        return;
+    }
+    double previous_beatpos = -1;
+    track::io::Beat beat;
+
+    foreach (double beatpos, beats) {
+        // beatpos is in frames. Do not accept fractional frames.
+        beatpos = floor(beatpos);
+        if (beatpos <= previous_beatpos || beatpos < 0) {
+            qDebug() << "kBeatMap::createFromVector: beats not in increasing order or negative";
+            qDebug() << "discarding beat " << beatpos;
+        } else {
+            beat.set_frame_position(beatpos);
+            m_beats.append(beat);
+            previous_beatpos = beatpos;
+        }
+    }
+    onBeatlistChanged();
+}
+
+int Beats::numBeatsInRangeNew(FrameNum startFrame, FrameNum endFrame) {
     double dLastCountedBeat = 0.0;
     int iBeatsCounter;
-    for (iBeatsCounter = 1; dLastCountedBeat < dEndSample; iBeatsCounter++) {
-        dLastCountedBeat = findNthBeat(dStartSample, iBeatsCounter);
+    for (iBeatsCounter = 1; dLastCountedBeat < endFrame; iBeatsCounter++) {
+        dLastCountedBeat = findNthBeat(startFrame, iBeatsCounter);
         if (dLastCountedBeat == -1) {
             break;
         }
     }
     return iBeatsCounter - 2;
 };
 
-double Beats::findNBeatsFromSample(double fromSample, double beats) const {
+QByteArray Beats::toByteArray() const {
+    QMutexLocker locker(&m_mutex);
+    // No guarantees BeatLists are made of a data type which located adjacent
+    // items in adjacent memory locations.
+    track::io::BeatMap map;
+
+    for (int i = 0; i < m_beats.size(); ++i) {
+        map.add_beat()->CopyFrom(m_beats[i]);
+    }
+
+    std::string output;
+    map.SerializeToString(&output);
+    return QByteArray(output.data(), output.length());
+}
+
+BeatsPointer Beats::clone() const {
+    QMutexLocker locker(&m_mutex);
+    // TODO(JVC)
+    BeatsPointer other(new Beats(*this));
+    return other;
+}
+
+QString Beats::getVersion() const {
```

Comment:
That this is called version is a bit confusing, It is more like a type. Do we have a strong demand to keep that wording or can we change it?